### PR TITLE
Feature/todo management

### PR DESCRIPTION
**Pull request by swarup-ranium**

### CODE REVIEW

File: tests/Browser/LoginTest.php [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    {
+        $this->browse(function (Browser $browser) {
+            $browser->visit('/login')
+                    ->type('email', 'swarup@ranium.com')
+                    ->type('password', '12345678')
+                    ->press('LOGIN')
+                    ->assertPathIs('/task');
+        });
+    }
+}
```
Comment (amit-ranium) [inline tests/Browser/LoginTest.php:68]: More tests needed. for ex, to check if a user is already logged in it should be redirected to the dashboard (or /task page here)

Comment (swarup-ranium) [reply tests/Browser/LoginTest.php:68]: ok sir

File: tests/Browser/CreateTaskTest.php [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+{
+    /**
+     * A Dusk test example.
+     *
+     * @return void
+     */
+    public function testExample()
+    {
+        $this->browse(function ($first) {
+            $first->loginAs(User::find(1))
```
Comment (amit-ranium) [inline tests/Browser/CreateTaskTest.php]: User object should be created separately

File: tests/Browser/LoginTest.php (hunk 1 above)
Comment (swarup-ranium) [reply tests/Browser/LoginTest.php:68]: You need to login for each test case, so can't check if the user is already logged in or not.

File: tests/Browser/CreateTaskCategoryTest.php [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+use Illuminate\Foundation\Testing\DatabaseMigrations;
+
+class CreateTaskCategoryTest extends DuskTestCase
+{
+    /**
+     * A Dusk test example.
+     *
+     * @return void
+     */
+    public function testExample()
```
Comment (abbasali) [inline tests/Browser/CreateTaskCategoryTest.php]: Why are all method named `testExample()`? The method name should be based on the thing we are testing in that method.  A better name in this case is `testCreateCategory()`

File: tests/Browser/CreateTaskCategoryTest.php [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    {
+        $this->browse(function ($first) {
+            $first->loginAs(User::find(1))
+                    ->visit('/taskCategory/create')
+                    ->assertTitle('Laravel')
+                    ->value('#name', 'Project #2')
+                    ->press('Add')
+                    ->assertPathIs('/taskCategory');
```
Comment (abbasali) [inline tests/Browser/CreateTaskCategoryTest.php]: You should be asserting the success message shown. Also assert that the newly added category appears on the listing page.

File: tests/Browser/CreateTaskCategoryTest.php [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+     *
+     * @return void
+     */
+    public function testExample()
+    {
+        $this->browse(function ($first) {
+            $first->loginAs(User::find(1))
+                    ->visit('/taskCategory/create')
+                    ->assertTitle('Laravel')
+                    ->value('#name', 'Project #2')
```
Comment (abbasali) [inline tests/Browser/CreateTaskCategoryTest.php]: Use Faker (WithFaker trait) to generate random name instead of hardcoding `Project #2`

File: tests/Browser/CreateTaskTest.php [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+            ->visit('/task/create')
+            ->assertTitle('Laravel')
+            ->value('#name', 'Task #2')
+            ->select('category_id', '1')
```
Comment (abbasali) [inline tests/Browser/CreateTaskTest.php]: Where is the category id coming from? You should be using a separate "test" database and it should be migrated for each test. See https://laravel.com/docs/6.x/dusk#migrations.

Also, use Factories to create data needed for test (in this case you will need to create categories using Factory before creating the task in dusk).

File: tests/Browser/ExampleTest.php [hunk 7]
```diff
@@ -0,0 +1,24 @@
+<?php
```
Comment (abbasali) [inline tests/Browser/ExampleTest.php]: Remove this example test case file.

File: tests/Browser/LoginTest.php (hunk 1 above)
Comment (abbasali) [reply tests/Browser/LoginTest.php:68]: There needs to be a test for negative scenario i.e. when you provide invalid login credentials. Assert whether you see an error message in that case and that the user is not logged in.

File: tests/Browser/RegisterTest.php [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+     *
+     * @return void
+     */
+    public function testExample()
+    {
+        $this->browse(function (Browser $browser) {
+            $browser->visit('/register')
+                    ->assertTitle('Laravel')
+                    // ->clickLink('Sign Up')
+                    ->value('#name', 'swarup manwatkar')
```
Comment (abbasali) [inline tests/Browser/RegisterTest.php]: Use Faker to generate name and all other values needed in the Form

**Review by abbasali — CHANGES_REQUESTED** [review]

File: tests/Browser/LoginTest.php (hunk 1 above)
Comment (swarup-ranium) [reply tests/Browser/LoginTest.php:68]: ok sir

File: tests/Browser/CreateTaskCategoryTest.php (hunk 3 above)
Comment (swarup-ranium) [reply tests/Browser/CreateTaskCategoryTest.php]: ok sir.

File: tests/Browser/CreateTaskCategoryTest.php (hunk 4 above)
Comment (swarup-ranium) [reply tests/Browser/CreateTaskCategoryTest.php]: ok sir

File: tests/Browser/CreateTaskCategoryTest.php (hunk 5 above)
Comment (swarup-ranium) [reply tests/Browser/CreateTaskCategoryTest.php]: ok sir.

File: tests/Browser/CreateTaskCategoryTest.php [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+            $first->loginAs($taskCategry->user_id)
+                    ->visit('/taskCategory/create')
+                    ->assertTitle('Laravel')
+                    ->value('#name', $taskCategry->name)
+                    ->press('Add')
+                    ->assertPathIs('/taskCategory')
+                    ->assertSee('Data Added successfully!')
+                    ->assertSee($taskCategry->name);
+        });
+    }
```
Comment (abbasali) [inline tests/Browser/CreateTaskCategoryTest.php:33]: There needs to be test cases for negative flow i.e. assert validation errors on leaving name empty etc. Same for other forms/models as well. Always there should be multiple cases covering both positive and negative flows a user can take while performing that action.

**Review by abbasali — CHANGES_REQUESTED** [review]

File: database/factories/TaskFactory.php [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+     *
+     * @return array
+     */
+    public function definition()
+    {
+        $taskCategory = TaskCategory::factory()->create();
+        return [
+            'name' => $this->faker->name,
+            'user_id' => $taskCategory->user_id,
+            'task_category_id' => $taskCategory->id,
```
Comment (abbasali) [inline database/factories/TaskFactory.php:30]: This should use `'task_category_id' => TaskCategory::factory(),`. Don't create task category directly on line 26.

File: tests/Browser/CreateTaskCategoryValidationTest.php [hunk 11]
```diff
@@ -0,0 +1,30 @@
+<?php
+
+namespace Tests\Browser;
+
+use Tests\DuskTestCase;
+use Laravel\Dusk\Browser;
+use App\Models\TaskCategory;
+use Illuminate\Foundation\Testing\DatabaseMigrations;
+
+class CreateTaskCategoryValidationTest extends DuskTestCase
```
Comment (abbasali) [inline tests/Browser/CreateTaskCategoryValidationTest.php]: No need to create a separate class for testing negative cases. All test cases related to Task Creation should go in a single class i.e. in `CreateTaskCategoryTest` class.

**Review by abbasali — CHANGES_REQUESTED** [review]

File: tests/Browser/CreateTaskCategoryValidationTest.php (hunk 11 above)
Comment (swarup-ranium) [reply tests/Browser/CreateTaskCategoryValidationTest.php]: Do I have to use if condition ?

File: tests/Browser/CreateTaskTest.php [hunk 12]
```diff
@@ -25,6 +25,9 @@ public function testCreateTask()
             ->value('#name', $task->name)
             ->select('category_id', $task->task_category_id)
             ->press('Add')
+            ->assertDontSee('The name field is required')
+            ->assertDontSee('The category id field is required')
+            ->assertDontSee('Whoops! Something went wrong!')
```
Comment (swarup-ranium) [inline tests/Browser/CreateTaskTest.php:30]: Added here

**Review by abbasali — APPROVED** [review]
